host/l2cap: fix sending error codes in connection requests

KKopyscinski opened the pull request:
Fixes issues related to Enhanced Credit Based connection request.

Fixes: #35840

KKopyscinski committed:
Bluetooth: Host: add missing zero CIDs in ecred_conn_req

Fix missing 0x0000 CIDs in response for ecred_conn_req. Previously,
when one of "All connections refused" error i variable was set to 0,
thus no CIDs were copied to response.

Signed-off-by: Krzysztof Kopyściński <krzysztof.kopyscinski@codecoup.pl>

KKopyscinski commented:
FYI @sjanc

KKopyscinski committed:
Bluetooth: Host: do not overwrite result in  le_ecred_conn_req

For more than one requested channels in le_ecred_conn_req only result is
overwritten in every iteration. This causes an issues if after failed
l2cap_chan_accept occures successful one: returned result will be
BT_L2CAP_LE_SUCCESS, where in reality should be returned error
and non-zero Destination CIDs will signal "Some connections refused".
Now, overwrite error only if the last one failed. If no errors occur,
result will remain BT_L2CAP_LE_SUCCESS.

This was affecting test cases L2CAP/ECFC/BV-20-C

Signed-off-by: Krzysztof Kopyściński <krzysztof.kopyscinski@codecoup.pl>

joerchan commented on subsys/bluetooth/host/l2cap.c:
```diff
@@ -1062,7 +1062,13 @@ static void le_conn_req(struct bt_l2cap *l2cap, uint8_t ident,
 	/* Check if connection has minimum required security level */
 	if (!l2cap_check_security(conn, server)) {
-		rsp->result = sys_cpu_to_le16(BT_L2CAP_LE_ERR_AUTHENTICATION);
+		if (server->sec_level == BT_SECURITY_L2) {
```
I think it would be better to have l2cap_check_security return the error code.

Also the error codes don't seem right. L2 -> ERR_ENCRYPTION, L3 -> ERR_AUTHENTICATION.

joerchan commented on subsys/bluetooth/host/l2cap.c:
```diff
@@ -1167,25 +1177,34 @@ static void le_ecred_conn_req(struct bt_l2cap *l2cap, uint8_t ident,
+		rsp->result = sys_cpu_to_le16(result);
+	}
 
 	bt_l2cap_send(conn, BT_L2CAP_CID_LE_SIG, buf);
```
There is a path now where buf could be uninitialized.

joerchan replied:
Or pointing to the original buffer received.

joerchan commented on subsys/bluetooth/host/l2cap.c:
```diff
@@ -1167,25 +1177,34 @@ static void le_ecred_conn_req(struct bt_l2cap *l2cap, uint8_t ident,
 	    result == BT_L2CAP_LE_ERR_PSM_NOT_SUPP ||
-	    result == BT_L2CAP_LE_ERR_AUTHENTICATION) {
-		memset(dcid, 0, sizeof(scid) * i);
+	    result == BT_L2CAP_LE_ERR_AUTHENTICATION ||
```
This code path seems very confusing. Is there any error codes that requires special handling here?
It would seem to be that this should be `if (result != BT_L2CAP_LE_ERR_SUCCESS)`.

KKopyscinski replied:
I wondered the same, so if there's no special cases I think it should be as you suggested

joerchan commented on subsys/bluetooth/host/l2cap.c:
```diff
@@ -1062,7 +1062,13 @@ static void le_conn_req(struct bt_l2cap *l2cap, uint8_t ident,
+			result =
+				sys_cpu_to_le16(BT_L2CAP_LE_ERR_AUTHORIZATION);
+		}
 		goto rsp;
```
rsp->result not initialized when sending the response.

KKopyscinski committed:
Bluetooth: Host: change error returned when ecred_conn_req

BT_L2CAP_LE_ERR_INVALID_PARAMS shall be returned instead of
BT_L2CAP_LE_ERR_UNACCEPT_PARAMS.

This was affecting test case L2CAP/ECFC/BV-26-C

Signed-off-by: Krzysztof Kopyściński <krzysztof.kopyscinski@codecoup.pl>

KKopyscinski commented on subsys/bluetooth/host/l2cap.c:
```diff
@@ -1158,10 +1158,9 @@ static void le_ecred_conn_req(struct bt_l2cap *l2cap, uint8_t ident,
-		result = l2cap_chan_accept(conn, server, scid, mtu, mps,
-					   credits, &chan[i]);
-		switch (result) {
+		result = MAX(rc = l2cap_chan_accept(conn, server, scid, mtu, mps,
```
@sjanc I agree, without MAX() macro it's much more readable. I updated the PR.

joerchan replied:
The other problem with the MAX would have been that l2cap_chan_accept would have been called twice.

carlescufi commented:
@Vudentz can you take a look?

joerchan commented on subsys/bluetooth/host/l2cap.c:
```diff
@@ -1003,14 +1004,26 @@ static uint16_t l2cap_chan_accept(struct bt_conn *conn,
+		if (server->sec_level == BT_SECURITY_L2) {
+			return BT_L2CAP_LE_ERR_ENCRYPTION;
+		} else if (server->sec_level == BT_SECURITY_L3) {
+			if (IS_ENABLED(CONFIG_BT_SMP_ENFORCE_MITM)) {
+				return BT_L2CAP_LE_ERR_AUTHORIZATION;
+			} else {
+				return BT_L2CAP_LE_ERR_AUTHENTICATION;
+			}
```
I don't think this part is right. Are there tests expecting BT_L2CAP_LE_ERR_AUTHORIZATION error code?

Authorization is asking the user for confirmation before continuing, returning it here based on ENFORCE_MITM is wrong.

KKopyscinski replied:
Tests expecting this error are L2CAP/LE/CFC/BV-13-C and L2CAP/ECFC/BV-13-C  (for ATT and EATT, respectively)

joerchan commented on subsys/bluetooth/host/l2cap.c:
```diff
@@ -1003,14 +1004,26 @@ static uint16_t l2cap_chan_accept(struct bt_conn *conn,
-	return conn->sec_level >= server->sec_level;
+	if (conn->sec_level < server->sec_level) {
+		if (server->sec_level == BT_SECURITY_L2) {
+			return BT_L2CAP_LE_ERR_ENCRYPTION;
```
From what I could tell from the spec then returning BT_L2CAP_LE_ERR_AUTHENTICATION was alright for this case as well.
```
Note: When encryption is not enabled, the result value “Connection refused –
insufficient authentication” does not indicate that MITM protection is required.
```

KKopyscinski replied:
Good point. I browsed through L2CAP API and this should be Authentication errors, as it was before commit. Authorization error should be returned if server accept callback returns -EACCES. I updated the PR

joerchan commented:
Could you update the commit message to use ":" instead of "/"
And preferable start with Bluetooth, so "Bluetooth: host: "

Vudentz commented on subsys/bluetooth/host/l2cap.c:
```diff
@@ -31,6 +31,7 @@
 
 #define L2CAP_LE_MIN_MTU		23
 #define L2CAP_ECRED_MIN_MTU		64
+#define L2CAP_ECRED_MAX_MTU		65533
```
Where is this value coming from, or is this UINT16_MAX - sizeof(uint16_t), afaik there is no such limitation since even if the very first 2 bytes are used for the SDU length that only means that an extra segment with 2 bytes would be needed if the MPS is UINT16_MAX.

KKopyscinski replied:
I browsed through config files some more, shouldn't it just be lower of the values BT_L2CAP_TX_MTU and BT_L2CAP_RX_MTU? Basically the same thing that is done with BT_ATT_MTU.

Vudentz commented on subsys/bluetooth/host/l2cap.c:
```diff
@@ -1167,27 +1175,34 @@ static void le_ecred_conn_req(struct bt_l2cap *l2cap, uint8_t ident,
-	rsp = net_buf_add(buf, sizeof(*rsp));
-	(void)memset(rsp, 0, sizeof(*rsp));
+		(void)memset(rsp, 0, sizeof(*rsp));
+		memset(dcid, 0, sizeof(scid));
```
Read the code comments above, the default is handling when some connections are refused so if you mean dcid to 0 any channel that did succeed would be set to 0.

Vudentz commented on subsys/bluetooth/host/l2cap.c:
```diff
@@ -1167,27 +1175,34 @@ static void le_ecred_conn_req(struct bt_l2cap *l2cap, uint8_t ident,
-	    result == BT_L2CAP_LE_ERR_PSM_NOT_SUPP ||
-	    result == BT_L2CAP_LE_ERR_AUTHENTICATION) {
-		memset(dcid, 0, sizeof(scid) * i);
+		net_buf_add_mem(buf, dcid, sizeof(scid));
```
This is also wrong, it should have been sizeof(scid) *i as the original code, although the memset above was probably wrong as well since it was reseting all dcids to 0 when in fact some may have been accepted.

KKopyscinski replied:
The problem I found is that if early error occurs (which is basically "All connections refused" case) i = 0, so no CIDs are sent back. I updated a PR

KKopyscinski commented:
Removed unnecessary error translation - it was already done in `l2cap_chan_accept()`

joerchan commented on subsys/bluetooth/host/l2cap.c:
```diff
@@ -1167,25 +1180,45 @@ static void le_ecred_conn_req(struct bt_l2cap *l2cap, uint8_t ident,
-		memset(dcid, 0, sizeof(scid) * i);
+			rsp = net_buf_add(buf, sizeof(*rsp));
+			(void)memset(rsp, 0, sizeof(*rsp));
+			memset(dcid, 0, sizeof(scid) * i);
```
This would be some connections refused then? memset of dcid to 0 seems wrong in this case. This is where we should send back the dcids allocated.

joerchan commented on subsys/bluetooth/host/l2cap.c:
```diff
@@ -1167,25 +1180,45 @@ static void le_ecred_conn_req(struct bt_l2cap *l2cap, uint8_t ident,
-	net_buf_add_mem(buf, dcid, sizeof(scid) * i);
+	if (i == 0) {
+		/* All connections refused */
+		net_buf_add_mem(buf, dcid, sizeof(scid));
```
If all connections are refused and we send only one entry then I think this would be easiest to handle by just having
`net_buf_add_le16(buf, 0);` and then get rid of the memset above of dcid. It just makes it harder to read the code, and memseting of dcid just to add a zero here is not very straightforward.

joerchan commented on subsys/bluetooth/host/l2cap.c:
```diff
@@ -1167,25 +1180,45 @@ static void le_ecred_conn_req(struct bt_l2cap *l2cap, uint8_t ident,
+			rsp->result = sys_cpu_to_le16(result);
+		}
 	} else if (ch) {
+		buf = l2cap_create_le_sig_pdu(buf, BT_L2CAP_ECRED_CONN_RSP, ident,
```
Failure to allocate buffer should be handled, in all cases.

joerchan commented on subsys/bluetooth/host/l2cap.c:
```diff
@@ -1167,25 +1180,45 @@ static void le_ecred_conn_req(struct bt_l2cap *l2cap, uint8_t ident,
 		rsp->mtu = sys_cpu_to_le16(ch->rx.mtu);
 		rsp->credits = sys_cpu_to_le16(ch->rx.init_credits);
+		rsp->result = sys_cpu_to_le16(result);
 	}
```
Since there is no else to this, it means that if none of the above happens we would use buf, which is pointing to the original request to send the response. That wouldn't be a good idea I think. Is there any way we could end up in a hypothetical else here?

KKopyscinski commented on subsys/bluetooth/host/l2cap.c:
```diff
@@ -1122,6 +1124,12 @@ static void le_ecred_conn_req(struct bt_l2cap *l2cap, uint8_t ident,
 
 	if (mtu < L2CAP_ECRED_MIN_MTU || mps < L2CAP_ECRED_MIN_MTU) {
+		BT_ERR("Invalid ecred conn req params");
+		result = BT_L2CAP_LE_ERR_INVALID_PARAMS;
```
I don't see tests for this MTU check for L2CAP without ECFC. Also, Core specification doesn't tell anything about these errors, only lists their codes. Maybe this can be added to `le_conn_req`, but it shouldn't affect any tests.

joerchan replied:
I was referring to this: https://github.com/zephyrproject-rtos/zephyr/blob/715b68d389b9cdb75208a48663035c48a34828b0/subsys/bluetooth/host/l2cap.c#L1044L1047

It is not related to your changes here, I just find it weird.

Do you know if there is any tests that verifies that behavior?

KKopyscinski replied:
Yes, that’s what I meant. I haven’t seen any tests of this, for not enhanced credit flow control

KKopyscinski commented on subsys/bluetooth/host/l2cap.c:
```diff
@@ -31,6 +31,8 @@
 
 #define L2CAP_LE_MIN_MTU		23
 #define L2CAP_ECRED_MIN_MTU		64
+#define L2CAP_ECRED_MAX_MTU		MIN(CONFIG_BT_L2CAP_TX_MTU, \
```
The thing with these L2CAP tests is that this EATT SPSM is used only as standard PSM for tests, but doesn't really test nor requires EATT. These tests can be executed with EATT disabled, and just BT_L2CAP_ECRED is required. In fact, this is how I test it. We just need to register server with EATT PSM and listen on it. There is test spec errata filed and accepted (TSE 16780), but TCLR is not yet released. There should be IXIT in PTS to set what PSM shall be used.

Second of all, yes, L2CAP/ECFC/BV-27-C was the reason for checking max MTU. I should've added this in commit message. I read through TS some more and I think there are multiple issues with this test: first, it requires to set this max MTU value in IXIT, but no such IXIT exists (only for ATT MTU). Secondly, the test passes even if no error is sent and L2CAP connection on EATT PSM is successful. What's more, because no value is set in IXIT, MTU in request is set to 0xFFFF, which should end test as inconclusive. I think this requires PTS errata.

To sum it up, I'll just remove this max MTU check and just keep  BT_L2CAP_LE_ERR_INVALID_PARAMS error.
I'll also file an PTS support case for L2CAP/ECFC/BV-27-C

joerchan replied:
Ok, thank you for the explanation. I agree with you that it is best to have this resolved with PTS first before we make a change here.
The thing that I couldn't understand from the test is why we would say the MTU is too large, since the MTU the remote is sending is the MTU that it can receive, it shouldn't be a problem that it is larger than the packets that we are sending.

carlescufi commented:
@KKopyscinski please rebase against latest main

KKopyscinski commented:
@carlescufi Rebased, but I'll return to this PR because it's not finished

joerchan commented on subsys/bluetooth/host/l2cap.c:
```diff
@@ -1167,9 +1168,13 @@ static void le_ecred_conn_req(struct bt_l2cap *l2cap, uint8_t ident,
-		result = l2cap_chan_accept(conn, server, scid, mtu, mps,
-					   credits, &chan[i]);
-		switch (result) {
+		rc = l2cap_chan_accept(conn, server, scid, mtu, mps,
```
The declaration of rc can be moved into this while loop.

joerchan commented on subsys/bluetooth/host/l2cap.c:
```diff
@@ -1159,6 +1159,7 @@ static void le_ecred_conn_req(struct bt_l2cap *l2cap, uint8_t ident,
 	/* Check if connection has minimum required security level */
 	if (!l2cap_check_security(conn, server)) {
 		result = BT_L2CAP_LE_ERR_AUTHENTICATION;
+
```
Revert blank line added here.

KKopyscinski commented:
@joerchan I updated the PR, now first commit has appropriate message (it fixes missing 0x0000 CIDs). I fixed the code accoriding to your revew. As of L2CAP/ECFC/BV-27-C, there is TSE15410, which confirms that MTU larger than something is nonsense, and it should be smaller. I think we should introduce some config setting for L2CAP_ECRED_MIN_MTU - it could be 64 for by default, could be set to any larger value. If MTU in req is smaller than 64 we shall return BT_L2CAP_LE_ERR_INVALID_PARAMS, and if smaller than L2CAP_ECRED_MIN_MTU - BT_L2CAP_LE_ERR_UNACCEPT_PARAMS. According to TSE, we will need L2CAP_ECRED_MIN_MTU larger than 64.

KKopyscinski commented:
Fixed Complience issues

KKopyscinski commented:
For L2CAP/ECFC/BV-26-C I submitted support request CASE0071314 (PTS doesn't see response)

joerchan reviewed (approved):
Thanks. Good work.

joerchan commented:
@sjanc Can you please revisit?

jhedberg commented:
Please link this to an issue that it fixes if this should be merged for 2.6. AFAIK only PRs that close issues are accepted at this late stage of the release process.

joerchan commented:
> Please link this to an issue that it fixes if this should be merged for 2.6. AFAIK only PRs that close issues are accepted at this late stage of the release process.

@jhedberg @KKopyscinski  I created the issue and added a reference.
